server/src/main/java/org/elasticsearch/index/query/functionscore/ScoreFunctionBuilder.java

```diff
@@ -70,10 +73,17 @@ public final void writeTo(StreamOutput out) throws IOException {
      */
     @SuppressWarnings("unchecked")
     public final FB setWeight(float weight) {
-        this.weight = weight;
+        this.weight = checkWeight(weight);
         return (FB) this;
     }
 
+    private Float checkWeight(Float weight) {
+        if (weight != null && weight < 0) {
+            DEPRECATION_LOGGER.deprecated("[weight] cannot be negative  for a filtering function [{}]", weight);
```

Comment:
Since this is master can you throw an `IllegalArgumentException` and we'll handle the deprecation when we backport to 6x.

Reply:
Hm... I thought that it is preferred to have one PR against `master` handing the deprecation (that is backported to 6.x) and a second (followup) PR for the removal (and IAE) that is only on `master`.
My bad;) I will change the PR shortly.

Reply:
For this one I think we can have two prs, one for the removal and one for the deprecation that targets 6x only. There are some [tests]https://github.com/elastic/elasticsearch/pull/33310/commits/62697443d9def3a2986ae2b177820646937868dd#diff-9af0a85c71293ca3d30f83fe80aedd28 in master that will fail in 6x because they produce negative weights so the backport is not trivial.